hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/protocol/ScmBlockLocationProtocolServerSideTranslatorPB.java

```diff
@@ -104,7 +104,7 @@ public ScmBlockLocationProtocolServerSideTranslatorPB(
   @Override
   public SCMBlockLocationResponse send(RpcController controller,
       SCMBlockLocationRequest request) throws ServiceException {
-    if (!scm.getScmContext().isLeader()) {
+    if (!scm.checkLeader()) {
```

Comment:
I don't know whether ratis now has the feature of leader-lease, with which scm will not need to check this every time.

Reply:
The reason behind adding this check is that read requests should not be processed before becoming leader becomes ready.

The reason is We do not want to serve the reads based on not up-to-date SCM DB.

Reply:
>>Should we separate NotLeaderException and LeaderNotReadyException ? client should failover for the former one, and retry the same SCM for the latter one
Good idea, let me update based on that.

Reply:
Should we separate NotLeaderException and LeaderNotReadyException ? client should failover for the former one, and retry the same SCM for the latter one.

As we have suggested Leader handling in Client, even though it says NotLeader Exception to the client, but it passes the leaderAddress in the case of LeaderNotReady it will be the same node address so retry will happen on the same Node. So, I believe we are good here.